Grant manage_threads to java.desktop for Tika

prdoyle opened the pull request:
### Problem

When processing PDFs, Apache Tika calls pdfbox, which creates a `MemoryCacheInputStream`, which (until JDK 26) uses `sun.java2d.Disposer`, whose static initializer creates a cleanup thread. Since `Disposer` is in the `java.desktop` module, which we exclude from system modules, the entitlement check for `manage_threads` fails.

### Workaround

For the time being, just have the server grant `manage_threads` to `java.desktop`. This is a fairly broad grant, so we'd like to find a more surgical solution, but this problem is observed in production right now, so we need a fix.

### Stack trace

```
org.elasticsearch.entitlement.runtime.api.NotEntitledException: component [(server)], module [java.desktop], class [class sun.java2d.Disposer], entitlement [manage_threads]
	at org.elasticsearch.entitlement@9.2.0/org.elasticsearch.entitlement.runtime.policy.PolicyCheckerImpl.notEntitled(PolicyCheckerImpl.java:467)
	at org.elasticsearch.entitlement@9.2.0/org.elasticsearch.entitlement.runtime.policy.PolicyCheckerImpl.checkFlagEntitlement(PolicyCheckerImpl.java:443)
	at org.elasticsearch.entitlement@9.2.0/org.elasticsearch.entitlement.runtime.policy.PolicyCheckerImpl.checkEntitlementPresent(PolicyCheckerImpl.java:481)
	at org.elasticsearch.entitlement@9.2.0/org.elasticsearch.entitlement.runtime.policy.PolicyCheckerImpl.checkManageThreadsEntitlement(PolicyCheckerImpl.java:406)
	at org.elasticsearch.entitlement@9.2.0/org.elasticsearch.entitlement.runtime.policy.ElasticsearchEntitlementChecker.check$java_lang_Thread$setDaemon(ElasticsearchEntitlementChecker.java:2647)
	at java.base/java.lang.Thread.setDaemon(Thread.java)
	at java.desktop/sun.java2d.Disposer.<clinit>(Disposer.java:80)
	at java.desktop/javax.imageio.stream.MemoryCacheImageInputStream.<init>(MemoryCacheImageInputStream.java:76)
	at org.apache.pdfbox.filter.LZWFilter.doLZWDecode(LZWFilter.java:78)
	at org.apache.pdfbox.filter.LZWFilter.decode(LZWFilter.java:70)
	at org.apache.pdfbox.filter.Filter.decode(Filter.java:97)
	at org.apache.pdfbox.filter.Filter.decode(Filter.java:285)
	at org.apache.pdfbox.cos.COSStream.createView(COSStream.java:196)
	at org.apache.pdfbox.pdmodel.PDPage.getContentsForRandomAccess(PDPage.java:267)
	at org.apache.pdfbox.pdmodel.PDPage.getContentsForStreamParsing(PDPage.java:256)
	at org.apache.pdfbox.pdfparser.PDFStreamParser.<init>(PDFStreamParser.java:59)
	at org.apache.pdfbox.contentstream.PDFStreamEngine.processStreamOperators(PDFStreamEngine.java:534)
	at org.apache.pdfbox.contentstream.PDFStreamEngine.processStream(PDFStreamEngine.java:515)
	at org.apache.pdfbox.contentstream.PDFStreamEngine.processPage(PDFStreamEngine.java:158)
	at org.apache.pdfbox.text.LegacyPDFStreamEngine.processPage(LegacyPDFStreamEngine.java:153)
	at org.apache.pdfbox.text.PDFTextStripper.processPage(PDFTextStripper.java:379)
	at org.apache.tika.parser.pdf.PDF2XHTML.processPage(PDF2XHTML.java:136)
	at org.apache.tika.parser.pdf.AbstractPDF2XHTML.processPages(AbstractPDF2XHTML.java:1362)
	at org.apache.pdfbox.text.PDFTextStripper.writeText(PDFTextStripper.java:252)
	at org.apache.tika.parser.pdf.PDF2XHTML.process(PDF2XHTML.java:107)
	at org.apache.tika.parser.pdf.PDFParser.parse(PDFParser.java:219)
	at org.apache.tika.parser.CompositeParser.parse(CompositeParser.java:298)
	at org.apache.tika.parser.AutoDetectParser.parse(AutoDetectParser.java:204)
	at org.apache.tika.Tika.parseToString(Tika.java:525)
	at org.elasticsearch.ingest.attachment.TikaImpl.parse(TikaImpl.java:73)
	at org.elasticsearch.ingest.attachment.AttachmentProcessor.execute(AttachmentProcessor.java:123)
	at org.elasticsearch.server@9.2.0/org.elasticsearch.ingest.CompoundProcessor.innerExecute(CompoundProcessor.java:171)
	at org.elasticsearch.server@9.2.0/org.elasticsearch.ingest.CompoundProcessor.execute(CompoundProcessor.java:146)
	at org.elasticsearch.server@9.2.0/org.elasticsearch.ingest.Pipeline.execute(Pipeline.java:214)
	at org.elasticsearch.server@9.2.0/org.elasticsearch.ingest.IngestDocument.executePipeline(IngestDocument.java:1125)
	at org.elasticsearch.server@9.2.0/org.elasticsearch.ingest.IngestService.executePipeline(IngestService.java:1369)
	at org.elasticsearch.server@9.2.0/org.elasticsearch.ingest.IngestService.executePipelines(IngestService.java:1199)
	at org.elasticsearch.server@9.2.0/org.elasticsearch.ingest.IngestService$1.doRun(IngestService.java:1039)
	at org.elasticsearch.server@9.2.0/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
	at org.elasticsearch.server@9.2.0/org.elasticsearch.ingest.IngestService.executeBulkRequest(IngestService.java:1046)
	at org.elasticsearch.server@9.2.0/org.elasticsearch.action.bulk.TransportAbstractBulkAction.processBulkIndexIngestRequest(TransportAbstractBulkAction.java:303)
	at org.elasticsearch.server@9.2.0/org.elasticsearch.action.bulk.TransportAbstractBulkAction.lambda$applyPipelines$3(TransportAbstractBulkAction.java:283)
	at org.elasticsearch.server@9.2.0/org.elasticsearch.action.ActionListener.run(ActionListener.java:465)
	at org.elasticsearch.server@9.2.0/org.elasticsearch.action.bulk.TransportAbstractBulkAction.applyPipelines(TransportAbstractBulkAction.java:273)
	at org.elasticsearch.server@9.2.0/org.elasticsearch.action.bulk.TransportAbstractBulkAction.applyPipelinesAndDoInternalExecute(TransportAbstractBulkAction.java:444)
	at org.elasticsearch.server@9.2.0/org.elasticsearch.action.bulk.TransportAbstractBulkAction$2.doRun(TransportAbstractBulkAction.java:190)
	at org.elasticsearch.server@9.2.0/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:1067)
	at org.elasticsearch.server@9.2.0/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1095)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:619)
	at java.base/java.lang.Thread.run(Thread.java:1447)
```

prdoyle committed:
Grant manage_threads to java.desktop for Tika

ldematte commented on test/framework/src/main/java/org/elasticsearch/bootstrap/TestScopeResolver.java:
```diff
@@ -24,14 +24,24 @@
+            // This case happens for JDK modules. Usually those are trivially allowed, but some are excluded,
+            // and those end up here.
+            // We have no test build info for those modules, so for now, let's just guess.
+            if (callerClass.getPackageName().equals("sun.java2d")) {
```
I understand why this needs to be a special case and reflect what happens in production, but I do wonder if we can come up with a more generic solution, so we don't have to keep adding things here.

What if we add a step/utility to our test framework, that inspects each module in `MODULES_EXCLUDED_FROM_SYSTEM_MODULES` and builds a map `packageName -> moduleName` for each package in each module, and / or `moduleName (+ component, always server) -> PolicyScope`? That would generalize this code and it would work with any package from any of these "JDK modules which are not in the system layer" modules.

It should not be too difficult, but maybe it's better to handle this in a follow-up. Wdyt?

prdoyle replied:
Sounds like a dynamic version of Test Build Info? That makes sense to me. We could also do it statically, actually, using the existing Gradle task, assuming we're able to know what JDK we're using for tests.

prdoyle committed:
Merge branch 'main' into disposer-entitlements

ldematte commented on libs/entitlement/src/main/java/org/elasticsearch/entitlement/bootstrap/HardcodedEntitlements.java:
```diff
@@ -114,7 +114,13 @@ private static List<Scope> createServerEntitlements(Path pidFile) {
+                "java.desktop",
+                List.of(
+                    new LoadNativeLibrariesEntitlement(),
+                    new ManageThreadsEntitlement() // For sun.java2d.Disposer
```
Nit: if you want, you can add a TODO here referencing https://elasticco.atlassian.net/browse/ES-12888 (which I just created)

ldematte commented on test/framework/src/main/java/org/elasticsearch/entitlement/runtime/policy/TestPolicyManager.java:
```diff
@@ -97,6 +97,10 @@ public final void clearModuleEntitlementsCache() {
 
     @Override
     protected boolean isTrustedSystemClass(Class<?> requestingClass) {
+        if (requestingClass.getPackageName().startsWith("sun.java2d")) {
```
Same here, this would become: `excludedSystemModulesByPackage.contains(requestingClass.getPackageName())`

ldematte reviewed (commented):
looks good, just some comments/questions

prdoyle committed:
Produce a PolicyScope for excluded system modules.

Fortunately, the JDK is still modular even if the tests are run sans modules,
so we can use the module API to identify the packages in every system module.

prdoyle committed:
TODO with Jira link

ldematte reviewed (approved):
Good job coding the `package -> scope` map for the excluded modules! I still sense  some friction (like it should be simpler), but I think this is the best we can do currently. LGTM
